src/awkward/_connect/numpy.py

```diff
@@ -141,9 +141,10 @@ def action(inputs, **ignore):
             return _array_ufunc_adjust(custom, inputs, kwargs, behavior)
 
         if ufunc is numpy.matmul:
-            custom_matmul = action_for_matmul(inputs)
-            if custom_matmul is not None:
-                return custom_matmul()
+            raise NotImplementedError
```

Comment:
> My understanding is that `return NotImplemented` will allow other arguments to implement `matmul`, e.g. an `ak.Array(...) @ np.array()` where the `ak.Array` is regular.

It would exit the function on `return` as it does on `raise`

Reply:
That's right, `return NotImplemented` lets other overloads take over, while `raise NotImplementedError` is a full-stop: no computation takes place. (These two cases unfortunately look very similar to one another.)

Doing a `return NotImplemented` would allow NumPy to take over, and it would try to cast the Awkward Array as a NumPy array by calling its `__array__` function. However, returning `NotImplemented` at this point in the code might not do it; we'd need to get the return value all the way back to the original `__array_ufunc__` call, if it needs to go through layers of the call stack in `broadcast_and_apply`, it might break something unrelated. To actually `return NotImplemented`, we might need to throw and catch an internal, custom exception, just so that we can skip most of the call stack and get the `NotImplemented` object directly out of `__array_ufunc__`.

Asking the question, "What _should_ we do?" I'm leaning more toward `raise NotImplementedError`, so that `@` always fails with an error message that indicates that this operator can and will be handled in the future; we might end up soliciting requests to implement it sooner (finding out just how many people need it). Doing `return NotImplemented` (properly) would make `@` work for regular arrays and not work for irregular arrays with a weird error message, most likely, "This array is not convertible to NumPy."

But in order for that to be useful, this PR should add an error message:

```python
            raise ak._errors.wrap_error(
                NotImplementedError(
                    "matrix multiplication (`@` or `np.matmul`) is not yet implemented for Awkward Arrays"
                )
            )
```

> I'm fairly sure that there is a mechanism like this, but I think it might only be for `__array_function__`.

Yes, the mechanism that returns `NotImplemented` to allow NumPy to try to cast the Awkward Array and use its own function is for NEP 18 (`__array_function__`), not NEP 13 (`__array_ufunc__`). With the exception of matrix multiplication, all ufuncs go through a common mechanism that applies to all Awkward Arrays except those with records (because we deliberately prevented it). None of the ufuncs (currently) `return NotImplemented` to let NumPy cast it.

Reply:
> None of the ufuncs (currently) `return NotImplemented` to let NumPy cast it.

This was my motivation when choosing `raise`.